File: Дорофеев/2476.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    void inorder(vector<int>& vec, TreeNode* root)
    {
        if (root == NULL)
        {
            return;
        }
        inorder(vec, root->left);
        vec.push_back(root->val);
        inorder(vec, root->right);
    }

    vector<vector<int>> closestNodes(TreeNode* root, vector<int>& queries)
    {

        vector<vector<int>> answer;
        vector<int> vec;
        inorder(vec, root);
        int n = vec.size();
        int mini, maxi;

        for (int q : queries)
        {
            mini = maxi = -1;
            //если элемент нашёлся, тогда он и мин, и макс
            if (binary_search(vec.begin(), vec.end(), q))
            {
                maxi = mini = q;
            }
            else
            {
                if (q >= vec[0])
                {
                    mini = lower_bound(vec.begin(), vec.end(), q) - vec.begin();
                    --mini;
                    if (mini >= 0)
                    {
                        mini = vec[mini];
                    }
                }
                if (q < vec[n - 1])
                {
                    maxi = upper_bound(vec.begin(), vec.end(), q) - vec.begin();
                    maxi = vec[maxi];
                }
            }

            answer.push_back({ mini,maxi });
        }

        return answer;
    }
};

File: Volokitin/2476.py
from bisect import bisect

class Solution:
    def closestNodes(self, root: Optional[TreeNode], queries: List[int]) -> List[List[int]]:
        array = []
        inorder_trav(root, array)
        answer = []
        for num in queries:
            answer.append(b_find(array, num))
        return answer

def inorder_trav(root, array):
    if root.left != None:
        inorder_trav(root.left, array)
    array.append(root.val)
    if root.right != None:
        inorder_trav(root.right, array)

def b_find(array, num):
    out = [0, 0]

    if num > array[-1]:
        out[0] = array[-1]
    if num < array[0]:
        out[0] = -1

    if num < array[0]:
        out[1] = array[0]
    if num > array[-1]:
        out[1] = -1
    
    if out[0] != 0 and out[1] != 0:
        return out
    
    if out[0] != 0 and out[1] == 0:
        out[1] = array[bisect_right(array, num)]
        return out
    if out[0] == 0 and out[1] != 0:
        out[0] = array[bisect_left(array, num)]
        return out
    
    i = bisect(array, num)
    if array[i - 1] == num:
        return [array[i - 1], array[i - 1]]
    else:
        return [array[i - 1], array[i]] 